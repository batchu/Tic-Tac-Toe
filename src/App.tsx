import * as React from 'react';
import './App.css';

interface SquareProps {
    value: string;
    onClick: () => void
}
class Square extends React.Component<SquareProps, {}> {
    render() {
        return (
            <button
                className="square"
                onClick={() => this.props.onClick()}
            >
                {this.props.value}
            </button>
        );
    }
}

interface BoardProps {

}

interface BoardState {
    squares: Array<string>
}

class Board extends React.Component<BoardProps, BoardState> {

    constructor() {
        super()
        this.state = {
            squares: Array(9).fill(null)
        }
    }

    handleClick(i: number) {
        const squares = this.state.squares.slice();
        squares[i] = 'X';
        this.setState({ 'squares': squares })
    }

    renderSquare(i: number) {
        return (
            <Square
                value={this.state.squares[i]}
                onClick={() => this.handleClick(i)}
            />
        );
    }

    render() {
        return (
            <div>
                <div className="status">Game Status</div>
                <div className="board-row">
                    {this.renderSquare(0)}
                    {this.renderSquare(1)}
                    {this.renderSquare(2)}
                </div>
                <div className="board-row">
                    {this.renderSquare(3)}
                    {this.renderSquare(4)}
                    {this.renderSquare(5)}
                </div>
                <div className="board-row">
                    {this.renderSquare(6)}
                    {this.renderSquare(7)}
                    {this.renderSquare(8)}
                </div>
            </div>
        );
    }
}

class Game extends React.Component {
    render() {
        return (
            <Board />
        );
    }
}

interface LinkProps {
    page: string
}

interface LinkState {
    class: STATUS
}

enum STATUS {
    NORMAL = 'NORMAL',
    HOVERED = 'HOVERED'
}

class Link extends React.Component<LinkProps, LinkState> {

    constructor(props: LinkProps) {
        super(props);

        this._onMouseEnter = this._onMouseEnter.bind(this);
        this._onMouseLeave = this._onMouseLeave.bind(this);

        this.state = {
            class: STATUS.NORMAL,
        };
    }

    _onMouseEnter() {
        this.setState({ class: STATUS.HOVERED });
    }

    _onMouseLeave() {
        this.setState({ class: STATUS.NORMAL });
    }

    render() {
        return (
            <a
                className={this.state.class}
                href={this.props.page || '#'}
                onMouseEnter={this._onMouseEnter}
                onMouseLeave={this._onMouseLeave}
            >
                {this.props.children}
            </a>
        );
    }

}

export default class App extends React.Component {
    render() {
        return (
            <Link page="https://batchu.net"> Batchu</Link>
        );
    }
}

export { Game }
export { Board }
export { Square }
export { Link }
