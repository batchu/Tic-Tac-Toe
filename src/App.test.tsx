import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as enzyme from 'enzyme';
import * as ReactTestRenderer  from 'react-test-renderer';
import * as Adapter from 'enzyme-adapter-react-16';

import  App from './App'
import { Board, Game, Square, Link} from './App';

test('App renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App/>, div);

});

test('Game renders properly', () => {
    const component = ReactTestRenderer.create(
        <Game/>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()

})

test.skip("Board's handleClick should update the squares array based on the value passed to it", () => {

    const instance: Board = new Board()
    instance.handleClick(5)
    expect(instance.state.squares[5]).toEqual('X')

})

test("When a Square is clicked, it should invoke the function attached to the click event", () => {

    const squareValue = 'X'
    let onClickwasCalled: boolean = false
    const onClick = () => {
        onClickwasCalled = true
    }
    enzyme.configure({adapter: new Adapter()})
    const instance = enzyme.shallow(<Square
        value={squareValue}
        onClick={onClick}
    />)

    expect(instance.find("button")).toEqual('X')
    expect(onClickwasCalled).toBeTruthy()

})

test('Link changes the class when hovered', () => {
    const component = ReactTestRenderer.create(
      <Link page="http://www.facebook.com">Facebook</Link>
    );
    let tree = component.toJSON();
  
    expect(tree).toMatchSnapshot();
  
    // manually trigger the callback
    if(tree!=null)
    tree.props.onMouseEnter();
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  
    // manually trigger the callback
    if(tree!=null)
    tree.props.onMouseLeave();
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

