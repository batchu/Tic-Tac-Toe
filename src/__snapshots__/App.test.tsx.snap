// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders properly 1`] = `
<div>
  <div
    className="status"
  >
    Game Status
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Link changes the class when hovered 1`] = `
<a
  className="NORMAL"
  href="http://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Facebook
</a>
`;

exports[`Link changes the class when hovered 2`] = `
<a
  className="HOVERED"
  href="http://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Facebook
</a>
`;

exports[`Link changes the class when hovered 3`] = `
<a
  className="NORMAL"
  href="http://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Facebook
</a>
`;
